File: backend/src/main/kotlin/com/alpha/domain/repository/EmailVerificationRepository.kt
package com.alpha.domain.repository

import com.alpha.domain.entity.EmailVerificationEntity
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Modifying
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param
import java.time.OffsetDateTime
import java.util.*

interface EmailVerificationRepository : JpaRepository<EmailVerificationEntity, UUID> {
    fun findByToken(token: String): EmailVerificationEntity?
    @Query("SELECT ev FROM EmailVerificationEntity ev WHERE ev.user.id = :userId")
    fun findByUserId(@Param("userId") userId: UUID): List<EmailVerificationEntity>
    
    @Query("SELECT ev FROM EmailVerificationEntity ev WHERE ev.expiresAt < :now")
    fun findExpiredVerifications(@Param("now") now: OffsetDateTime): List<EmailVerificationEntity>
    
    @Modifying
    @Query("DELETE FROM EmailVerificationEntity ev WHERE ev.expiresAt < :now")
    fun deleteExpiredVerifications(@Param("now") now: OffsetDateTime): Int
    
    @Modifying
    @Query("DELETE FROM EmailVerificationEntity ev WHERE ev.user.id = :userId")
    fun deleteByUserId(@Param("userId") userId: UUID): Int
}

File: backend-scala-zio/src/main/scala/com/alpha/repository/EmailVerificationRepository.scala
package com.alpha.repository

import zio.*
import com.alpha.domain.model.*
import java.util.UUID
import java.time.OffsetDateTime
import io.getquill.*
import io.getquill.extras.OffsetDateTimeOps

trait EmailVerificationRepository:
  def findById(id: UUID): Task[Option[EmailVerification]]
  def findByUserId(userId: UUID): Task[List[EmailVerification]]
  def findByToken(token: String): Task[Option[EmailVerification]]
  def findActiveByUserId(userId: UUID): Task[Option[EmailVerification]]
  def create(verification: EmailVerification): Task[UUID]
  def markAsUsed(id: UUID): Task[Int]
  def deleteExpired: Task[Int]

object EmailVerificationRepository:
  val layer: ZLayer[PostgresCtx, Nothing, EmailVerificationRepository] =
    ZLayer.fromFunction(new EmailVerificationRepositoryImpl(_))

class EmailVerificationRepositoryImpl(ctx: PostgresCtx) extends EmailVerificationRepository:

  import ctx.*

  override def findById(id: UUID): Task[Option[EmailVerification]] =
    ZIO.attempt:
      run(query[EmailVerification].filter(_.id == lift(id))).headOption

  override def findByUserId(userId: UUID): Task[List[EmailVerification]] =
    ZIO.attempt:
      run(query[EmailVerification].filter(_.userId == lift(userId))).toList

  override def findByToken(token: String): Task[Option[EmailVerification]] =
    ZIO.attempt:
      run(query[EmailVerification].filter(_.token == lift(token))).headOption

  override def findActiveByUserId(userId: UUID): Task[Option[EmailVerification]] =
    ZIO.attempt:
      run(query[EmailVerification]
        .filter(v =>
          v.userId == lift(userId) && v.isUsed == lift(false) && v.expiresAt > lift(
            java.time.OffsetDateTime.now()))).headOption

  override def create(verification: EmailVerification): Task[UUID] =
    ZIO.attempt:
      run(query[EmailVerification].insertValue(lift(verification)))
      verification.id

  override def markAsUsed(id: UUID): Task[Int] =
    ZIO.attempt:
      run(query[EmailVerification]
        .filter(_.id == lift(id))
        .update(_.isUsed -> lift(true)))
      1

  override def deleteExpired: Task[Int] =
    ZIO.attempt:
      run(query[EmailVerification]
        .filter(_.expiresAt < lift(java.time.OffsetDateTime.now()))
        .delete)
      0
